File: app/src/main/java/com/ryanrvldo/moviecatalogue/viewmodel/SeasonsViewModel.java
package com.ryanrvldo.moviecatalogue.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ryanrvldo.moviecatalogue.data.Repository;
import com.ryanrvldo.moviecatalogue.data.source.model.Season;

public class SeasonsViewModel extends ViewModel {
    private LiveData<Season> seasons;

    private Repository repository;

    public SeasonsViewModel(Repository repository) {
        this.repository = repository;
    }

    private void setSeasons(int tvId, int seasonNumber) {
        seasons = new MutableLiveData<>();
        seasons = repository.getSeasonDetail(tvId, seasonNumber);
    }

    public LiveData<Season> getSeasons(int tvId, int seasonNumber) {
        if (seasons == null) {
            setSeasons(tvId, seasonNumber);
        }
        return seasons;
    }
}

File: app/src/main/java/com/ryanrvldo/moviecatalogue/ui/viewmodel/SeasonsViewModel.kt
package com.ryanrvldo.moviecatalogue.ui.viewmodel

import androidx.hilt.Assisted
import androidx.hilt.lifecycle.ViewModelInject
import androidx.lifecycle.LiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import com.ryanrvldo.moviecatalogue.data.Repository
import com.ryanrvldo.moviecatalogue.data.model.Season

class SeasonsViewModel @ViewModelInject constructor(
    private val repository: Repository,
    @Assisted private val savedStateHandle: SavedStateHandle
) : ViewModel() {

    private var tvShowId = savedStateHandle.get<Int>(TV_ID_KEY) ?: 0
    private var seasonNumber = savedStateHandle.get<Int>(SEASONS_KEY) ?: 0

    val seasons: LiveData<Season> by lazy {
        repository.getSeasonDetails(tvShowId, seasonNumber)
    }

    companion object {
        const val TAG = "SeasonsVM"
        const val TV_ID_KEY = "TV_SHOW_ID"
        const val SEASONS_KEY = "SEASON_NUMBER"
    }
}